lib/mdl/rules.rb

```diff
@@ -796,3 +796,85 @@ def check_blockquote(errors, elements)
     error_lines
   end
 end
+
+rule 'MD055', 'Tables should be well-formed' do
+  tags :tables
+  aliases 'table-format-validation'
+  check do |doc|
+    error_lines = []
+    tables = doc.find_type_elements(:table)
+    lines = doc.lines
+    num_lines = lines.length
+
+    tables.each do |table|
+      loc = table.options[:location]-1
+      header = lines[loc]
+
+      if header[0] != '|' || header[-1] != '|' || header.length == 1
+        # Tables should start and end with '|'
+        error_lines << (loc+1)
+      end
+
+      headings = header.strip().split('|').select{ |x| x.length > 0 }
+      num_headings = headings.length
+
+      if loc+1 < num_lines
+        # Second row should be in the form | ---- | ---- |
+        second_row = lines[loc+1]
+        columns = second_row.strip().split('|').select{ |x| x.length > 0 }
+        if num_headings != columns.length
+          # Number of headings do not match with the second row
+          error_lines << (loc+2)
+        end
+        
+        # This pattern matches if
+        #   1) The row starts and stops with | characters
+        #   2) Only consists of characters '|', '-', ':' and whitespace
+        #   3) Each section between the seperators (i.e. '|')
+        #      a) has at least three consecutive dashes
+        #      b) can have whitespace at the beginning or the end
+        #      c) can have colon before and/or after the dashes (for alignment purposes)
+        #   Some examples:
+        #       |-----|----|-------|      --> matches
+        #       |:---:|:---|-------|      --> matches
+        #       |  :------:  | ----|      --> matches
+        #       | - - - | - - - |         --> does NOT match
+        #       |::---|                   --> does NOT match
+        #       |----:|:--|----|          --> does NOT match
+        pattern = /^(\|\s*:?-{3,}:?\s*)+\|$/
+        unless second_row.match(pattern)
+          # Second row is not in the form described by the pattern
+          error_lines << (loc+2)
+        end
+
+        # Check next lines until the end of the table
+        line_num = loc+2
+        while second_row.strip() != '' && line_num < num_lines
```

Comment:
`second_row` doesn't change, it doesn't make sense to compare it in each iteration of the loop, much less to `strip` it in each iteration of the loop.

Perhaps you want something like:

```
unless second_row.strip.empty?
  line_num = loc + 1
  while line_num < num_lines
    ...
  end
end
```

?

(You can also play with something like `(num_lines - (loc + 2)).times do |num|`, if that appeals to you.
Also, what's a case where the second line is blank, but there's more lines? We can probably skill the whole `unless` above, I think?